### PR TITLE
Reshaping and Joins

### PR DESCRIPTION
**Pull request by RohitRathore1**

### CODE REVIEW

**Comment by RohitRathore1** [comment]
It is against #2789 and ready for review.

**Comment by bkamins** [comment]
So should #2789 be closed now?

File: docs/src/assets/readme.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
+# Introduction
```
Comment (bkamins) [inline docs/src/assets/readme.md:1]: do not create the second file with the same contents as README.md in this directory

**Comment by RohitRathore1** [comment]
> So should #2789 be closed now?

Yes.

File: docs/src/assets/risk.csv [hunk 2]
```diff
@@ -0,0 +1,1001 @@
+id,Risk
```
Comment (bkamins) [inline docs/src/assets/risk.csv:1]: what is the line ending you use in this file?

File: docs/src/man/reshaping_new.md [hunk 3]
```diff
@@ -0,0 +1,498 @@
+# Reshaping Data
+
+*DataFrames.jl* is a popular julia library for tabular data which makes easier to read and transform data.
```
Comment (bkamins) [inline docs/src/man/reshaping_new.md:3]: why is this needed? I thought it would be in the middle of the tutorial.

File: docs/src/man/reshaping_new.md [hunk 4]
```diff
@@ -0,0 +1,498 @@
...
+It provides the abstractions of data frames and series, similar to those in *R*, and *Pandas*.
+
+In *DataFrames.jl* data reshaping means the transformaton of the structure of the table or vector (for example 
+data frames or series) to make it suitable for further analysis. Let us assume we have a data frame with 
+multiindices on the rows and columns. In `DataFrames.jl` currently we have two functions: `stack`, and `unstack`
```
Comment (bkamins) [inline docs/src/man/reshaping_new.md:8]: Where did you get this part from? In DataFrames.jl we do not support such functionalities.

File: docs/src/man/reshaping_new.md [hunk 5]
```diff
@@ -0,0 +1,498 @@
...
+that we can use for reshaping our data. Their goals are very simple:
+- `stack` allows us to go from wide to long data format;
+- `unstack` allows us to go from long to wide data format.
```
Comment (bkamins) [inline docs/src/man/reshaping_new.md:11]: maybe explain what wide and long mean or give a link that explains it.

File: docs/src/man/reshaping_new.md [hunk 6]
```diff
@@ -0,0 +1,498 @@
...
+In *DataFrames.jl* data reshaping means the transformaton of the structure of the table or vector (for example 
+data frames or series) to make it suitable for further analysis. Let us assume we have a data frame with 
+multiindices on the rows and columns. In `DataFrames.jl` currently we have two functions: `stack`, and `unstack`
+that we can use for reshaping our data. Their goals are very simple:
+- `stack` allows us to go from wide to long data format;
+- `unstack` allows us to go from long to wide data format. 
+
+!!! note
+
+  Parts of this section has been taken from [Bogumił Kamiński](https://bkamins.github.io/julialang/2021/05/28/pivot.html)
```
Comment (bkamins) [inline docs/src/man/reshaping_new.md:15]: you can remove it.

File: docs/src/man/reshaping_new.md [hunk 7]
```diff
@@ -0,0 +1,498 @@
...
+
+First we will have a look that how to create multiple indices. There is a simplest way to create a multi index
```
Comment (bkamins) [inline docs/src/man/reshaping_new.md:17]: DataFrames.jl does not support indices, nor multi indices. You can do `groupby` instead.

File: docs/src/man/reshaping_new.md [hunk 8]
```diff
@@ -0,0 +1,498 @@
...
+
+First we will have a look that how to create multiple indices. There is a simplest way to create a multi index 
+data frame object is by passing a list of two or more arrays to the constructor. Have a look to the below example:
+
+```jldoctest dataframe
+julia> using DataFrames
+
+julia> grocessary = DataFrame(year = repeat(2019:2021, inner=4),
+                                list = repeat(["Sugar", "Soap", "Fruits", "Oil"], outer=3),
+                                month=1:12, Date=15:26)
```
Comment (bkamins) [inline docs/src/man/reshaping_new.md:25]: what is the interpretation of column `:Date`?

File: docs/src/man/reshaping_new.md [hunk 9]
```diff
@@ -0,0 +1,498 @@
...
+12×4 DataFrame
+ Row │ year   list    month  Date
+     │ Int64  String  Int64  Int64
+─────┼─────────────────────────────
+   1 │  2019  Sugar       1     15
```
Comment (bkamins) [inline docs/src/man/reshaping_new.md:30]: it seems that the data should rather be quaterly not monthly.

File: docs/src/man/reshaping_new.md [hunk 10]
```diff
@@ -0,0 +1,498 @@
...
+   5 │  2020  Sugar       5     19
+   6 │  2020  Soap        6     20
+   7 │  2020  Fruits      7     21
+   8 │  2020  Oil         8     22
+   9 │  2021  Sugar       9     23
+  10 │  2021  Soap       10     24
+  11 │  2021  Fruits     11     25
+  12 │  2021  Oil        12     26
+
+julia> costs = select(grocessary, :year, :list, [:month, :Date] .=> x -> x/2,
```
Comment (bkamins) [inline docs/src/man/reshaping_new.md:43]: why is this needed?

File: docs/src/man/reshaping_new.md [hunk 11]
```diff
@@ -0,0 +1,498 @@
...
+   6 │  2020  Soap        3.0     10.0
+   7 │  2020  Fruits      3.5     10.5
+   8 │  2020  Oil         4.0     11.0
+   9 │  2021  Sugar       4.5     11.5
+  10 │  2021  Soap        5.0     12.0
+  11 │  2021  Fruits      5.5     12.5
+  12 │  2021  Oil         6.0     13.0
+```
+
+Now before moving further let's practice a bit basic data transformation skills:
```
Comment (bkamins) [inline docs/src/man/reshaping_new.md:63]: this comment is not clear. I thought the point of this section is to discuss `stack`.

File: docs/src/man/reshaping_new.md [hunk 12]
```diff
@@ -0,0 +1,498 @@
...
+
+```jldoctest dataframe
+julia> long_grocessary = stack(grocessary, [:month, :Date], [:year, :list],
+                          variable_name=:days, value_name=:grocessary)
```
Comment (bkamins) [inline docs/src/man/reshaping_new.md:67]: what does `:grocessary` mean? Is it a typo?

File: docs/src/man/reshaping_new.md [hunk 13]
```diff
@@ -0,0 +1,498 @@
...
+   9 │  2021  Sugar       4.5     11.5
+  10 │  2021  Soap        5.0     12.0
+  11 │  2021  Fruits      5.5     12.5
+  12 │  2021  Oil         6.0     13.0
+```
+
+Now before moving further let's practice a bit basic data transformation skills:
+
+```jldoctest dataframe
+julia> long_grocessary = stack(grocessary, [:month, :Date], [:year, :list],
```
Comment (bkamins) [inline docs/src/man/reshaping_new.md:66]: I think it is important to explain what you are doing here.

File: docs/src/man/reshaping_new.md [hunk 14]
```diff
@@ -0,0 +1,498 @@
...
+  18 │  2020  Soap    Date       10.0
+  19 │  2020  Fruits  Date       10.5
+  20 │  2020  Oil     Date       11.0
+  21 │  2021  Sugar   Date       11.5
+  22 │  2021  Soap    Date       12.0
+  23 │  2021  Fruits  Date       12.5
+  24 │  2021  Oil     Date       13.0
+                        9 rows omitted
+
+julia> long = innerjoin(long_grocessary, long_costs, on=[:year, :list, :days])
```
Comment (bkamins) [inline docs/src/man/reshaping_new.md:114]: will section on joins be earler or later in the tutorial?

File: docs/src/man/reshaping_new.md [hunk 15]
```diff
@@ -0,0 +1,498 @@
...
+24×5 DataFrame
+ Row │ year   list    days    grocessary  costs
```
Comment (bkamins) [inline docs/src/man/reshaping_new.md:116]: it is not clear for me - from the interpretation point of view - why `:days` combines both `month` and `Date`

File: docs/src/man/reshaping_new.md [hunk 16]
```diff
@@ -0,0 +1,498 @@
...
+  24 │  2021  Oil     Date            26     13.0
+                                    9 rows omitted
+```                                           
+
+## The basics of `stack`
+
+The concept of stacking comes in handy when we have data with multi indices. We use the `stack()` function 
+to reshape the data frame by converting the data into the stacked form. Reshaping the data from wide to 
+long format using the `stack` function means moving (also rotating or pivoting) the innermost column index 
+to become the innermost row index.
```
Comment (bkamins) [inline docs/src/man/reshaping_new.md:143]: It seems that this is copied from pandas. We do not have such concepts in DataFrames.jl

File: docs/src/assets/risk.csv (hunk 2 above)
Comment (RohitRathore1) [reply docs/src/assets/risk.csv:1]: Line ending?

File: docs/src/man/reshaping_new.md (hunk 14 above)
Comment (RohitRathore1) [reply docs/src/man/reshaping_new.md]: It will be earlier.

File: docs/src/assets/risk.csv (hunk 2 above)
Comment (bkamins) [reply docs/src/assets/risk.csv:1]: You have created german.csv with CRLF line endings. The rest of the files in the package uses LF line endings. I have checked myself. This file also uses CRLF endings.

It is OK to leave this this way. We just need to remember that we have such a situation.

If you do not know much about CRLF vs LF issue here is a good read: https://www.aleksandrhovhannisyan.com/blog/crlf-vs-lf-normalizing-line-endings-in-git/

**Comment by bkamins** [comment]
Hi - are you planning to work on this PR?

**Comment by RohitRathore1** [comment]
> Hi - are you planning to work on this PR?

Yes, can you review this once?

File: docs/make.jl [hunk 17]
```diff
@@ -21,6 +21,8 @@ makedocs(
     pages = Any[
         "Introduction" => "index.md",
         "First Steps with DataFrames.jl" => "man/basics.md",
+        "Joins" => "man/joins_new.md",
```
Comment (bkamins) [inline docs/make.jl:24]: this structure does not seem to be good.
Can you please make the "First Steps with DataFrames.jl" a top level tutorial page, and then have sections as separate pages in it (actually already https://dataframes.juliadata.org/stable/man/basics/ is probably too long and should be split into 2 separate pages; but this can be done later in a separate PR)

File: docs/src/assets/readme.md (hunk 1 above)
Comment (bkamins) [reply docs/src/assets/readme.md:1]: The comment still applies.

Comment (bkamins) [reply docs/src/assets/readme.md:1]: Just add the comment about the Risk column to the already existing file.

File: docs/src/man/joins_new.md [hunk 18]
```diff
@@ -0,0 +1,879 @@
...
+
+    Some of the part has been taken from [Bence Komarniczky](https://towardsdatascience.com/joining-dataframes-in-julia-c435e3da32f3)
+
+Sound familiar? You may have come across this question plenty of times on online discussion
+forums. Working with one data frame is fairly straightforward but things become challenging when
+we have data spread across two or more data frames. This is where the concept of *joins* come in.
+
+*Joins* is a very common and important operation that arises in the world of tabulated data. 
+A join across two data frames is the action of combining the two datasets based on shared column values 
+that exist across them. We call this column (or columns) the *key(s)*. So, each record from the
```
Comment (bkamins) [inline docs/src/man/joins_new.md:16]: we can keep keys and records terminology, but we do not use it in DataFrames.jl documentation. Where have you taken it from?

File: docs/src/man/joins_new.md [hunk 19]
```diff
@@ -0,0 +1,879 @@
...
+*Joins* is a very common and important operation that arises in the world of tabulated data. 
+A join across two data frames is the action of combining the two datasets based on shared column values 
+that exist across them. We call this column (or columns) the *key(s)*. So, each record from the 
+first data frame is matched to a record in the second data frame — as long as the records’ key values are 
+the same. 
+
+Previously, we have used German Credit Card dataset. Now let's say you are working
+on a classification problem but we do not have target column in our dataset. We are given two data frames – 
+one which contains data about customer and the other that has taget (risk factor) information. We will 
+use these data frames  to understand how the different types of *joins* work using `DataFrames.jl`.
```
Comment (bkamins) [inline docs/src/man/joins_new.md:23]:
```suggestion
use these data frames  to understand how the different types of *joins* work using DataFrames.jl.
```

File: docs/src/man/joins_new.md [hunk 20]
```diff
@@ -0,0 +1,879 @@
...
+
+For example, suppose that we have the following two data frames:
+
+```jldoctest dataframe
+julia> using DataFrames
+
+julia> using CSV
+
+julia> german_ref = CSV.read(joinpath(dirname(pathof(DataFrames)),
+                                       "..", "docs", "src", "assets", "german.csv"),
```
Comment (bkamins) [inline docs/src/man/joins_new.md:33]: misaligned indent

File: docs/src/man/joins_new.md [hunk 21]
```diff
@@ -0,0 +1,879 @@
...
+   5 │     4  bad
+   6 │     5  good
+   7 │     6  good
+   8 │     7  good
+   9 │     8  good
+  10 │     9  bad
+```
+
+We now have 2 data frames:
+- `german_ref` holds the `:id`s, `:Age`, `:Sex`, `:Job` and `:Housing`
```
Comment (bkamins) [inline docs/src/man/joins_new.md:100]: this data frame holds also other columns

File: docs/src/man/joins_new.md [hunk 22]
```diff
@@ -0,0 +1,879 @@
...
+- `risk_ref` holds the `:id`s and `:Risk`.
+
+In relational database theory, to combine two datasets referred to as a join. The columns used to
```
Comment (bkamins) [inline docs/src/man/joins_new.md:103]: you have already said this - why repeating?

File: docs/src/man/joins_new.md [hunk 23]
```diff
@@ -0,0 +1,879 @@
...
+
+The following functions are provided to perform seven kinds of joins:
+
+- `innerjoin`: the output contains rows for values of the key that exist in all passed data frames.
+- `leftjoin`: the output contains rows for values of the key that exist in the first (left) argument, 
+  whether or not that value exists in the second (right) argument.
+- `rightjoin`: the output contains rows for values of the key that exist in the second (right) argument, 
+  whether or not that value exists in the first (left) argument.
+- `outerjoin`: the output contains rows for values of the key that exist in any of the passed data frames.
+- `semijoin`: Like an inner join, but output is restricted to columns from the first (left) argument.
```
Comment (bkamins) [inline docs/src/man/joins_new.md:114]: where have you taken this description from? We use another definition based on subset or rows(which is not 100% equivalent)

File: docs/src/man/joins_new.md [hunk 24]
```diff
@@ -0,0 +1,879 @@
...
+
+## Inner Join
+
+The output contains rows for values of the key that exist in both the first (left) and second (right) 
+arguments to join.
+
+We would like to combine the 2 data frames so that we can see columns of `german_ref` and `risk_ref` together. 
+We do a `innerjoin`!
+
+Rules for the `on` keyword argument are as follows:
```
Comment (bkamins) [inline docs/src/man/joins_new.md:128]: it is not clear what `on` keyword argument does from your earlier description.

File: docs/src/man/joins_new.md [hunk 25]
```diff
@@ -0,0 +1,879 @@
...
+- a single `Symbol` or string if joining on one column with the same name, e.g. `on=:id`
+- a `Pair` of `Symbol`s or strings if joining on one column with different names, e.g. `on=:id=>:id2`
+- a `vector` of `Symbol`s or strings if joining on multiple columns with the same name, 
+  e.g. `on=[:id1, :id2]`
+- a `vector` of `Pair`s of `Symbol`s or strings if joining on multiple columns with the same name,
```
Comment (bkamins) [inline docs/src/man/joins_new.md:133]: incorrect description

File: docs/src/man/joins_new.md [hunk 26]
```diff
@@ -0,0 +1,879 @@
...
+- a single `Symbol` or string if joining on one column with the same name, e.g. `on=:id`
+- a `Pair` of `Symbol`s or strings if joining on one column with different names, e.g. `on=:id=>:id2`
+- a `vector` of `Symbol`s or strings if joining on multiple columns with the same name, 
+  e.g. `on=[:id1, :id2]`
+- a `vector` of `Pair`s of `Symbol`s or strings if joining on multiple columns with the same name, 
+  e.g. `on=[:a1=>:a2, :b1=>:b2]`
+- a `vector` containing a combination of `Symbol`s or strings or `Pair` of Symbols or strings, 
+  e.g. `on=[:a1=>:a2, :b1]`
+
+Here, we are working with a average (in size) data set that contains both the `:Age` and `:Sex` for each `:id`.
```
Comment (bkamins) [inline docs/src/man/joins_new.md:138]: what do you mean by average here?

File: docs/src/man/joins_new.md [hunk 27]
```diff
@@ -0,0 +1,879 @@
...
+   second (right) argument. As with semijoin, output is restricted to columns from the first (left) argument.
+- `crossjoin`: The output is the cartesian product of rows from all passed data frames.
+
+
+## Inner Join
+
+The output contains rows for values of the key that exist in both the first (left) and second (right) 
+arguments to join.
+
+We would like to combine the 2 data frames so that we can see columns of `german_ref` and `risk_ref` together.
```
Comment (bkamins) [inline docs/src/man/joins_new.md:125]: where do you join these tables below?

File: docs/src/man/joins_new.md [hunk 28]
```diff
@@ -0,0 +1,879 @@
...
+   4 │     3     45  male        2  free     good
+   5 │     4     53  male        2  free     bad
+   6 │     5     35  male        1  free     good
+   7 │     6     53  male        2  own      good
+   8 │     7     35  male        3  rent     good
+   9 │     8     61  male        1  own      good
+  10 │     9     28  male        3  own      bad                                          
+```
+
+Let’s go through this in detail. Arguments 1 and 2 `german_ref` & `risk_ref` are the two tables we’re
```
Comment (bkamins) [inline docs/src/man/joins_new.md:159]: the description is inconsistent with the code.

File: docs/src/man/joins_new.md [hunk 29]
```diff
@@ -0,0 +1,879 @@
...
+   8 │     7     35  male        3  rent     good
+   9 │     8     61  male        1  own      good
+  10 │     9     28  male        3  own      bad                                          
+```
+
+Let’s go through this in detail. Arguments 1 and 2 `german_ref` & `risk_ref` are the two tables we’re 
+joining. Argument 3 `on` tells us what the key column is. We’ll use this column to match the observations 
+across the data frames. As you can see, we ended up with *1000 rows* and *11 columns*. Now go back to the 
+beginning and check out how the 2 original data frames looked like. Make sure you understand why we ended up 
+with these rows only. Since we’re using `innerjoin`, we’re only keeping `:id`s that appear in both data frames —
```
Comment (bkamins) [inline docs/src/man/joins_new.md:163]: this is not evident from the example as you have kept all the rows actually

File: docs/src/man/joins_new.md [hunk 30]
```diff
@@ -0,0 +1,879 @@
...
+   5 │     8     61  male        1  own      good
+   6 │     9     28  male        3  own      bad
+   7 │     0     67  male        2  own      missing
+   8 │     1     22  female      2  own      missing
+   9 │     2     49  male        1  own      missing
+  10 │     3     45  male        2  free     missing
+```
+
+Here we kept all of the observations from data frame `german` no matter what is going on in data frame `risk1`. 
+For records without a match in data frame `german`, the `:Risk` column has `missingas` value. This makes sense
```
Comment (bkamins) [inline docs/src/man/joins_new.md:215]: typo in missing

File: docs/src/man/joins_new.md [hunk 31]
```diff
@@ -0,0 +1,879 @@
...
+The above four joins make up the basics of data frame merging. We can remember like this:
+- `innerjoin`: The output contains rows for values of the key that exist in both the first (left) and 
+  second (right) arguments to `join`.
+- `leftjoin`: The output contains rows for values of the key that exist in the first (left) argument to 
+  `join`, whether or not that value exists in the second (right) argument.
+- `rightjoin`: The output contains rows for values of the key that exist in the second (right) argument 
+  to `join`, whether or not that value exists in the first (left) argument.
+- `outerjoin`: The output contains rows for values of the key that exist in the first (left) or second 
+  (right) argument to `join`.
+
```
Comment (bkamins) [inline docs/src/man/joins_new.md:289]: also mention at the end what happens in case of duplicates in `on` column and how non-standard values in `on column are handled (most importantly `missing`)

File: docs/src/man/joins_new.md [hunk 32]
```diff
@@ -0,0 +1,879 @@
...
+─────┼──────────────────────────────────────
+   1 │     4     53  male        2  free
+   2 │     5     35  male        1  free
+   3 │     6     53  male        2  own
+   4 │     7     35  male        3  rent
+   5 │     8     61  male        1  own
+   6 │     9     28  male        3  own
+```
+
+The below example will return true, demonstrating that a `semijoin` is the same as `innerjoin` with
```
Comment (bkamins) [inline docs/src/man/joins_new.md:332]: as commented above - this is not true in general.

File: docs/src/man/joins_new.md [hunk 33]
```diff
@@ -0,0 +1,879 @@
...
+```
+
+## Cross Join
+
+The output is the cartesian product of rows from the first (left) and second (right) arguments to join.
+
+Now, let's have a look on `crossjoin`:
+
+```jldoctest dataframe
+julia> crossjoin(german, risk, makeunique=true);
```
Comment (bkamins) [inline docs/src/man/joins_new.md:347]: no output

File: docs/src/man/joins_new.md [hunk 34]
```diff
@@ -0,0 +1,879 @@
...
+```
+
+In the above example, A `crossjoin` took all the rows from `german` data frame and for each row, it matches it 
+to every single row of `risk` data frame. At first impression, this might not make any sense, but it’s a great 
+method for finding all combinations of 2 data frames. Our new data frame has 100 rows.
```
Comment (bkamins) [inline docs/src/man/joins_new.md:352]: better use smaller data frames and show the output

File: docs/src/man/joins_new.md [hunk 35]
```diff
@@ -0,0 +1,879 @@
...
+
+!!! note
+
+    Cross joins are the only kind of join that does not use a `on` key:
+    `crossjoin(persons, earnings, makeunique = true)`
```
Comment (bkamins) [inline docs/src/man/joins_new.md:357]: why repeating this with some random data frame names?

File: docs/src/man/joins_new.md [hunk 36]
```diff
@@ -0,0 +1,879 @@
...
+
+For more clarification, now let’s say we want to design a new data frame by changing the ingredients. To
```
Comment (bkamins) [inline docs/src/man/joins_new.md:359]: Clarification of what?

Maybe also just add a comment here where these examples are taken from?

File: docs/src/man/joins_new.md [hunk 37]
```diff
@@ -0,0 +1,879 @@
...
+   3 │ marshmallow             0.03
+    
+julia> coating_layer = DataFrame(coating = ["caramel", "chocolate sauce"],
+                                 coating_cost = [0.01, 0.05]) # table of coatings
+2×2 DataFrame
+ Row │ coating          coating_cost
+     │ String           Float64
+─────┼───────────────────────────────
+   1 │ caramel                  0.01
+   2 │ chocolate sauce          0.05
```
Comment (bkamins) [inline docs/src/man/joins_new.md:380]: maybe shorten the strings so that the final table fits in the output?

File: docs/src/man/joins_new.md [hunk 38]
```diff
@@ -0,0 +1,879 @@
...
+   2 │ biscuit                 0.05  chocolate sauce          0.05        0.1
+   3 │ chocolate biscuit       0.08  caramel                  0.01        0.09
+   4 │ chocolate biscuit       0.08  chocolate sauce          0.05        0.13
+   5 │ marshmallow             0.03  caramel                  0.01        0.04 ⋯
+   6 │ marshmallow             0.03  chocolate sauce          0.05        0.08
+```
+
+## Multiple keys to join on
+
+Now that we have better chocolate, let’s learn how to `join` on multiple columns. Extending the
```
Comment (bkamins) [inline docs/src/man/joins_new.md:418]: what you do mean by "chocolate" here?

File: docs/src/man/joins_new.md [hunk 39]
```diff
@@ -0,0 +1,879 @@
...
+   3 │     2  good
+   4 │     3  good
+   5 │     4  bad
+   6 │     5  good
+   7 │     6  good
+   8 │     7  good
+   9 │     8  good
+  10 │     9  bad
+
+julia> D_earnings.city = ["Orai", "Gwalior"][mod1.(1:10, 2)]
```
Comment (bkamins) [inline docs/src/man/joins_new.md:469]: same here

File: docs/src/man/joins_new.md [hunk 40]
```diff
@@ -0,0 +1,879 @@
...
+   7 │     6  good    Orai
+   8 │     7  good    Gwalior
+   9 │     8  good    Orai
+  10 │     9  bad     Gwalior
+```
+
+One way we can think of this is that we have 2 separate data frames. One in Orai and another Gwalior. 
+With the systems not knowing about each other, they keep track of the users’ `:id`s separately. So the 
+name of user 1 in Orai is not the same as user 1 in Gwalior. Indeed they are different users! So 
+when we merge the 2 data frames we want to make sure that the names and earnings are not only matched in
```
Comment (bkamins) [inline docs/src/man/joins_new.md:518]: we do not see this in effect in your example

File: docs/src/man/joins_new.md [hunk 41]
```diff
@@ -0,0 +1,879 @@
...
+  we treat `missing` as any other value without giving it a special treatment;
+- `:notequal` (a new option): in this case `missing` is considered to be not equal to any other value
+  (including `missing`).
+
+But there are some consequences of the `:notequal` rule. In `innerjoin` this means that rows with `missing`
+values will be dropped both in left and right table. In `leftjoin`, `semijoin`, and `antijoin` they are 
+dropped from the right table only (which means that if `missing` is present in the left table it is retained
+in processing but considered not to match any row in right table). Similarly in `rightjoin` rows with `missing`
+are dropped from left table only. The case that is most difficult to handle is `outerjoin`. The reason is that
+if `missing` would be present in both left and right table they would be considered not equal and produced
```
Comment (bkamins) [inline docs/src/man/joins_new.md:714]: in `outerjoin` we do not support `:notequal`

File: docs/src/man/joins_new.md [hunk 42]
```diff
@@ -0,0 +1,879 @@
...
+  11 │    19  missing          missing           good    right_only
+  12 │    20  missing          missing           good    right_only
+  13 │    21  missing          missing           good    right_only
+  14 │    22  missing          missing           good    right_only
+  15 │    23  missing          missing           good    right_only
+  16 │    24  missing          missing           good    right_only
+  17 │    25  missing          missing           good    right_only
+```
+
+For `leftjoin` and `rightjoin` we retained `missing` but only in the table for which all
```
Comment (bkamins) [inline docs/src/man/joins_new.md:836]: all this descriptions are incorrect and the behavior you see is unrelated with `matchmissing` kwarg value.

File: docs/src/man/reshaping_new.md [hunk 43]
```diff
@@ -0,0 +1,337 @@
...
+
+In `DataFrames.jl` currently we have two functions: `stack`, and `unstack`
+that we can use for reshaping our data. Their goals are very simple:
+- `stack` allows us to go from wide to long data format;
+- `unstack` allows us to go from long to wide data format. 
+
+In wide data it has a column for each variable whereas long format data has 
+a column for possible variable types & a column for the values of those variables.
+
+Reshape data from wide to long format using the stack function:
```
Comment (bkamins) [inline docs/src/man/reshaping_new.md:11]: this sentence does not match what you do below

File: docs/src/man/reshaping_new.md [hunk 44]
```diff
@@ -0,0 +1,337 @@
...
+   6 │     5     35  male        1  free     missing
+   7 │     6     53  male        2  own      quite rich
+   8 │     7     35  male        3  rent     little
+   9 │     8     61  male        1  own      rich
+  10 │     9     28  male        3  own      little                                                  
+```
+
+## The basics of `stack`
+
+Stack return the long-format `DataFrame` with columns for each of the ID variables, column
```
Comment (bkamins) [inline docs/src/man/reshaping_new.md:62]: it is not clear what ID variables are

File: docs/src/man/reshaping_new.md [hunk 45]
```diff
@@ -0,0 +1,337 @@
...
+```
+
+## The basics of `stack`
+
+Stack return the long-format `DataFrame` with columns for each of the ID variables, column 
+value name holding the values of the stacked columns, and column variable name a vector holding 
+the name of the corresponding measurement variables.
+
+Here, `german` is the `AbstractDataFrame` which has to be stacked and the columns to
+be stacked are `:Age`, `:Job`, `:Housing`, `:Sex` which is known as the measurement variables;
```
Comment (bkamins) [inline docs/src/man/reshaping_new.md:67]: description is inconsistent with code

File: docs/src/man/reshaping_new.md [hunk 46]
```diff
@@ -0,0 +1,337 @@
...
+`:variable` is the name of our new stacked columns which are holding the names of each of the
+measurement variables, and `value` is the name of the new stacked columns containing the values
+from each of the measurement variables:
+
+```jldoctest dataframe
+julia> stack(german, [:Age, :Job, :Housing])
```
Comment (bkamins) [inline docs/src/man/reshaping_new.md:73]: I would recommend you to start with examples showing all arguments passed explicitly

File: docs/src/man/reshaping_new.md [hunk 47]
```diff
@@ -0,0 +1,337 @@
...
+  25 │ male    Housing   free
+  26 │ male    Housing   free
+  27 │ male    Housing   own
+  28 │ male    Housing   rent
+  29 │ male    Housing   own
+  30 │ male    Housing   own
+                15 rows omitted
+```
+
+Below we have omitted second argument in `stack`, so all other columns are assumed to be the
```
Comment (bkamins) [inline docs/src/man/reshaping_new.md:120]: you did the same above

File: docs/src/man/reshaping_new.md [hunk 48]
```diff
@@ -0,0 +1,337 @@
...
+[id_variables](https://github.com/JuliaData/DataFrames.jl/blob/f690aa49e958f51e0c3c579b6def1f11be214d98/src/abstractdataframe/reshape.jl#:~:text=id_vars%60%20%3A%20the%20identifier,are%20not%20%60measure_vars%60):
```
Comment (bkamins) [inline docs/src/man/reshaping_new.md:121]: do not use such links as they will become outdated

File: docs/src/man/reshaping_new.md [hunk 49]
```diff
@@ -0,0 +1,337 @@
...
+  37 │     6  male    Saving accounts  quite rich
+  38 │     7  male    Saving accounts  little
+  39 │     8  male    Saving accounts  rich
+  40 │     9  male    Saving accounts  little
+                                   25 rows omitted
+```                                           
+
+## The basics of `unstack`
+
+*Unstack* converts the data frame `german` from it long to wide format. Row and column
```
Comment (bkamins) [inline docs/src/man/reshaping_new.md:177]: you do not unstack `german` below. Also the order of column keys is not guaranteed to be in order of appearance in general

File: docs/src/man/reshaping_new.md [hunk 50]
```diff
@@ -0,0 +1,337 @@
...
+  35 │     4  little           Housing   free
+  36 │     5  missing          Housing   free
+  37 │     6  quite rich       Housing   own
+  38 │     7  little           Housing   rent
+  39 │     8  rich             Housing   own
+  40 │     9  little           Housing   own
+                                25 rows omitted
+```
+
+We can even skip passing the `:variable` and `:value` values as positional arguments, as they
```
Comment (bkamins) [inline docs/src/man/reshaping_new.md:205]: as commented above - better start with the example where all arguments are passed.

File: docs/src/man/reshaping_new.md [hunk 51]
```diff
@@ -0,0 +1,337 @@
...
+   4 │     3  little           45   male    2    free
+   5 │     4  little           53   male    2    free
+   6 │     5  missing          35   male    1    free
+   7 │     6  quite rich       53   male    2    own
+   8 │     7  little           35   male    3    rent
+   9 │     8  rich             61   male    1    own
+  10 │     9  little           28   male    3    own
+```
+
+If the remaining columns are unique, we can skip the id variable and use:
```
Comment (bkamins) [inline docs/src/man/reshaping_new.md:227]: what do you mean by "remaining columns are unique"?

File: docs/src/man/reshaping_new.md [hunk 52]
```diff
@@ -0,0 +1,337 @@
...
+   4 │     3  little           45   male    2    free
+   5 │     4  little           53   male    2    free
+   6 │     5  missing          35   male    1    free
+   7 │     6  quite rich       53   male    2    own
+   8 │     7  little           35   male    3    rent
+   9 │     8  rich             61   male    1    own
+  10 │     9  little           28   male    3    own
+```
+
+We can rename the unstacked columns:
```
Comment (bkamins) [inline docs/src/man/reshaping_new.md:283]: maybe add more explanation?

File: docs/src/man/reshaping_new.md [hunk 53]
```diff
@@ -0,0 +1,337 @@
...
+```
+
+### Pivot tables with `unstack`
+
+Most likely we want to aggregate grocessary per year using the `sum` function. This is a classic pivot 
+table task. In DataFrames.jl currently one does it in two steps: first aggregate, then reshape. Here 
+is how we can do it (We are showing two separate steps, but you could use e.g. [Chain.jl](https://github.com/jkrumbiegel/Chain.jl) to streamline the processing):
+
+```jldoctest dataframe
+julia> df = combine(groupby(german, [:Age, :Sex]), :Job => sum => :Job)
```
Comment (bkamins) [inline docs/src/man/reshaping_new.md:310]: what sense does it make to sum `:Job`?

**Comment by bkamins** [comment]
I have finished initial comments. In general for the next review I would recommend you to:
1. carefully read the rendered HTML and make sure that text and code are consistent.
2. make sure that your examples are not only technically correct but also have a sensible interpretation.

Thank you for working on this.

**Comment by bkamins** [comment]
I am closing this PR due to inactivity (there is a risk that things get outdated). Please open a new updated PR if you would like to submit your proposal to change the documentation. Thank you!
